src/test/txrequest_tests.cpp

```diff
@@ -0,0 +1,539 @@
+            if ((config >> 3) == 3) { // A reponse will arrive for the transaction
+                scenario.ReceivedResponse(peer, gtxid);
+                scenario.Check(peer, {}, 0, 0, 0, "s10");
+                return;
+            }
+        }
+    }
+
+    scenario.AdvanceTime(RandomTime());
+    if (config & 4) { // The peer will go offline
+        scenario.DisconnectedPeer(peer);
+    } else { // The transaction is no longer needed
+        scenario.ForgetTxHash(gtxid.GetHash());
+    }
+    scenario.Check(peer, {}, 0, 0, 0, "s11");
+}
+
+/** Add to scenario a test with a single tx announced by two peers, to verify the
+ *  right peer is selected for requests.
+ *
+ * config is an integer between 0 and 31, which controls which variant of the test is used.
+ */
+void BuildPriorityTest(Scenario& scenario, int config)
+{
+    scenario.SetTestName(strprintf("Priority(config=%i)", config));
+
+    // Three peers. They will announce in order.
+    auto peer1 = scenario.NewPeer(), peer2 = scenario.NewPeer();
+    // Construct a transaction that under random rules would be preferred by peer2 or peer1,
+    // depending on configuration.
+    bool prio1 = config & 1;
+    auto gtxid = prio1 ? scenario.NewGTxid({{peer1, peer2}}) : scenario.NewGTxid({{peer2, peer1}});
+    bool pref1 = config & 2, pref2 = config & 4;
+
+    scenario.ReceivedInv(peer1, gtxid, pref1, MIN_TIME);
+    scenario.Check(peer1, {gtxid}, 1, 0, 0, "p1");
+    if (InsecureRandBool()) scenario.AdvanceTime(RandomTime());
+    scenario.Check(peer1, {gtxid}, 1, 0, 0, "p2");
+
+    scenario.ReceivedInv(peer2, gtxid, pref2, MIN_TIME);
+    bool stage2_prio =
+        // At this point, peer2 will be given priority if:
+        // - It is preferred and peer1 is not
+        (pref2 && !pref1) ||
+        // - They're in the same preference class,
+        //   and the randomized priority favors peer2 over peer1.
+        (pref1 == pref2 && !prio1);
+    uint64_t priopeer = stage2_prio ? peer2 : peer1, otherpeer = stage2_prio ? peer1 : peer2;
+    scenario.Check(otherpeer, {}, 1, 0, 0, "p3");
+    scenario.Check(priopeer, {gtxid}, 1, 0, 0, "p4");
+    if (InsecureRandBool()) scenario.AdvanceTime(RandomTime());
+    scenario.Check(otherpeer, {}, 1, 0, 0, "p5");
+    scenario.Check(priopeer, {gtxid}, 1, 0, 0, "p6");
+
+    // We possibly request from the selected peer.
+    if (config & 8) {
+        scenario.RequestedTx(priopeer, gtxid, MAX_TIME);
+        scenario.Check(priopeer, {}, 0, 1, 0, "p7");
+        scenario.Check(otherpeer, {}, 1, 0, 0, "p8");
+        if (InsecureRandBool()) scenario.AdvanceTime(RandomTime());
+    }
+
+    // The peer which was selected (or requested from) now goes offline, or a NOTFOUND is received from them.
+    if (config & 16) {
+        scenario.DisconnectedPeer(priopeer);
+    } else {
+        scenario.ReceivedResponse(priopeer, gtxid);
+    }
+    if (InsecureRandBool()) scenario.AdvanceTime(RandomTime());
```

Comment:
I think advancing time should be a config alternative of its own, otherwise can you dissociate the state transition REQUESTED -> COMPLETED triggered by a `ReceivedResponse/DisconnectPeer` from a `GetRequestable` one ?

Reply:
Turned it into a config bit.